### PR TITLE
[IMP] web, *: Stop using url fragment for state of the page

### PR DESCRIPTION
**Pull request by jpp-odoo**
Before this commit, we used url fragment (#) to store the state of
the page in the form of pairs of key/values.

The aim of this commit, is to be more coherent with the web practices.
The fragment is traditionally used to indicate the id of the HTML
element (an anchor) on the page that the browser should scroll to, this
is automatically done by the browser.

This commit, change the current practice and start using the
query/search string of the url to store the state of the page.

Part of task-id : 3557575

### CODE REVIEW

File: addons/web/static/src/core/browser/router_service.js [hunk 1]
```diff
@@ -103,47 +103,46 @@ export function parseSearchQuery(search) {
...
+    return { pathname, search: { ...searchQuery, ...hashQuery } };
 }
 
 function makeRouter(env) {
     const bus = env.bus;
     const lockedKeys = new Set();
     let current = getRoute(browser.location);
     let pushTimeout;
-    browser.addEventListener("hashchange", (ev) => {
+    browser.addEventListener("searchchange", (ev) => {
```
Comment (sdegueldre) [inline addons/web/static/src/core/browser/router_service.js:121]: maybe we shouldn't dispatch artificial events on window?

File: doc/cla/individual/pallavisrivastavaa.md
Comment (rfr-odoo) [inline doc/cla/individual/pallavisrivastavaa.md:1]: Remove CLA ?

File: addons/web/static/src/webclient/clickbot/clickbot.js [hunk 2]
```diff
@@ -15,46 +15,77 @@ const BLACKLISTED_MENUS = [
...
+                testedMenus: [],
+                testedFilters: 0,
+                testedModals: 0,
+                appIndex: 0,
+                menuIndex: 0,
+                subMenuIndex: 0,
+            },
+            stateHandler
+        );
+    }
```
Comment (sdegueldre) [inline addons/web/static/src/webclient/clickbot/clickbot.js:83]:
```js
const state = reactive({}, () => browser.localStorage.setItem("running.clickbot", JSON.stringify(state)))
browser.localStorage.setItem("running.clickbot", JSON.stringify(state));
```

File: doc/cla/individual/pallavisrivastavaa.md
Comment (jpp-odoo) [reply doc/cla/individual/pallavisrivastavaa.md:1]: I think I delete a wrong file :-)

File: addons/web/static/src/webclient/clickbot/clickbot_loader.js [hunk 3]
```diff
@@ -20,6 +21,11 @@ export function runClickTestItem({ env }) {
     };
 }
 
+const runningClickbot = JSON.parse(browser.localStorage.getItem("running.clickbot")) || {};
```
Comment (jpp-odoo) [inline addons/web/static/src/webclient/clickbot/clickbot_loader.js:24]: state

File: addons/web/static/src/webclient/clickbot/clickbot.js [hunk 4]
```diff
@@ -438,39 +479,31 @@ async function testApp(element) {
...
-            if (!app) {
-                throw new Error(`No app found for xmlid ${xmlId}`);
-            }
-            await testApp(app);
+            state.app = xmlId;
+            await testApp(state.app);
         } else {
-            while ((app = await getNextApp())) {
-                await testApp(app);
+            while ((state.app = await getNextApp(restartApp))) {
```
Comment (jpp-odoo) [inline addons/web/static/src/webclient/clickbot/clickbot.js:493]: if (state.app)
 testApp

remove restartApp

File: addons/web/static/src/webclient/clickbot/clickbot.js [hunk 5]
```diff
@@ -416,18 +438,37 @@ async function testMenuItem(element) {
...
- * @param {DomElement} element: the App menu item
+ * @param {appXmlId} xmlID: the App menu item
  * @returns {Promise}
  */
-async function testApp(element) {
-    browser.console.log(`Testing app menu: ${element.dataset.menuXmlid}`);
-    testedApps.push(element.dataset.menuXmlid);
-    await testMenuItem(element);
-    if (appsMenusOnly === true) {
+async function testApp(appXmlId) {
```
Comment (jpp-odoo) [inline addons/web/static/src/webclient/clickbot/clickbot.js:444]: appXmlId not needed we can catch it from the state

File: addons/web/static/src/webclient/clickbot/clickbot.js [hunk 6]
```diff
@@ -238,18 +255,23 @@ async function getNextMenu() {
...
  */
-async function getNextApp() {
-    let apps;
-    if (isEnterprise) {
-        await ensureHomeMenu();
-        apps = document.querySelectorAll(".o_apps .o_app");
-    } else {
-        await ensureAppsMenu();
-        apps = document.querySelectorAll(".o_navbar_apps_menu .dropdown-item");
+async function getNextApp(currentApp) {
```
Comment (jpp-odoo) [inline addons/web/static/src/webclient/clickbot/clickbot.js:258]: currentApp not needed anymore

File: addons/web/static/src/core/browser/router_service.js [hunk 7]
```diff
@@ -103,61 +111,62 @@ export function parseSearchQuery(search) {
...
+    return route.pathname + (search ? "?" + search : "");
 }
 
 function getRoute(urlObj) {
     const { pathname, search, hash } = urlObj;
     const searchQuery = parseSearchQuery(search);
-    const hashQuery = parseHash(hash);
-    return { pathname, search: searchQuery, hash: hashQuery };
+    const hashQuery = parseHash(hash); // For retro-compatibility purpose
+    return { pathname, search: { ...searchQuery, ...hashQuery } };
```
Comment (jpp-odoo) [inline addons/web/static/src/core/browser/router_service.js:121]: return { ...searchQuery }
Find another way to do the retro-compatibility, to allow it to re-write the url

File: addons/web/static/src/core/browser/router_service.js [hunk 8]
```diff
@@ -34,48 +36,54 @@ function parseString(str) {
...
 
-function computeNewRoute(hash, replace, currentRoute) {
+function computeNewRoute(search, replace, currentRoute) {
     if (!replace) {
-        hash = Object.assign({}, currentRoute.hash, hash);
+        search = Object.assign({}, currentRoute.search, search);
+    } else {
+        for (const key of PROTECTED_KEYS) {
+            if (key in currentRoute.search) {
+                search[key] = currentRoute.search[key];
```
Comment (jpp-odoo) [inline addons/web/static/src/core/browser/router_service.js:73]: if (!search[key])

File: addons/website/static/src/client_actions/website_preview/website_preview.js [hunk 9]
```diff
@@ -181,13 +181,13 @@ export class WebsitePreview extends Component {
...
             // When reaching a "regular" url of the webclient's router, an
-            // hashchange event should be dispatched to properly display the
+            // PopStateEvent event should be dispatched to properly display the
             // content of the previous URL before reaching the client action,
             // which was lost after being replaced for the frontend's URL.
             const handleBackNavigation = () => {
                 if (window.location.pathname === '/web') {
-                    window.dispatchEvent(new HashChangeEvent('hashchange', {
-                        newURL: window.location.href.toString()
+                    window.dispatchEvent(new PopStateEvent('popstate', {
```
Comment (aab-odoo) [inline addons/website/static/src/client_actions/website_preview/website_preview.js:189]: test but probably no longer necessary

File: addons/website_sale/static/src/js/website_sale.js [hunk 10]
```diff
@@ -82,9 +82,11 @@ export const WebsiteSale = publicWidget.Widget.extend(VariantMixin, cartHandlerM
 
         this._startZoom();
 
-        window.addEventListener('hashchange', () => {
-            this._applyHash();
-            this.triggerVariantChange(this.$el);
+        window.addEventListener('popstate', (ev) => {
```
Comment (aab-odoo) [inline addons/website_sale/static/src/js/website_sale.js:85]: check

File: addons/website/static/src/client_actions/website_preview/website_preview.js [hunk 11]
```diff
@@ -145,8 +145,8 @@ export class WebsitePreview extends Component {
 
         onWillUnmount(() => {
             this.websiteService.context.showResourceEditor = false;
-            const { pathname, search, hash } = this.iframe.el.contentWindow.location;
-            this.websiteService.lastUrl = `${pathname}${search}${hash}`;
+            const { pathname, search } = this.iframe.el.contentWindow.location;
+            this.websiteService.lastUrl = `${pathname}${search}`;
```
Comment (robinlej) [inline addons/website/static/src/client_actions/website_preview/website_preview.js:149]: I don't think we want to remove the hash here, this one could be legit as we're taking the iframe's location, not the one of the window.

**Review by robinlej — COMMENTED** [review]
General comments:
- when hovering over links in the navbar, they are still shown with the hash (from website, it's the only thing that appears; from the Odoo backend everything seems to be translated into a search query, but the # remains).

- probably linked: there are many occurrences of `web#...` in the code (not only in website). Are these to be changed in a later commit?

**Comment by Julien00859** [comment]
Just to speak about the terminologie, the [RFC-3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3) compliant name for "hash string" is actually "fragment". Can you add "fragment" next to "hash string" inside the title because to me it wasn't clear what "hash string" was about at first :p

**Review by aab-odoo — APPROVED** [review]

File: addons/web/static/src/core/anchor_scroll_prevention.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+/** @odoo-module **/
+
+import { EventBus } from "@odoo/owl";
+import { browser } from "./browser/browser";
+
+export const ScrollerBus = new EventBus();
+
+browser.addEventListener("click", (ev) => {
+    const href = ev.target.closest("a")?.getAttribute("href");
```
Comment (kebeclibre) [inline addons/web/static/src/core/anchor_scroll_prevention.js:9]: may be useful to be in capture mode.
An event stopped may trigger its default anyway, and this handler would not see it

File: addons/web/static/src/core/browser/router.js [hunk 13]
```diff
@@ -36,53 +36,49 @@ function parseString(str) {
...
-        Object.entries(hash)
-            .filter(([, v]) => v !== undefined)
+        Object.entries(obj)
+            .filter(([, v]) => v !== valueToRemove)
             .map(([k, v]) => [k, cast(v)])
     );
 }
 
+function sanitizeSearch(search) {
+    return sanitize(search);
```
Comment (kebeclibre) [inline addons/web/static/src/core/browser/router.js:75]: I'd explicitly set `undefined` for clarity, but it is only my very personal feeling

Comment (jpp-odoo) [reply addons/web/static/src/core/browser/router.js:75]: https://github.com/odoo/odoo/pull/146070#discussion_r1448531831

**Comment by aab-odoo** [comment]
robodoo rebase-ff r+

**Comment by aab-odoo** [comment]
robodoo r+

**Pull request closed without merging**